### ISSUE TITLE
[micronaut-tomcat-jdbc] "Table 'test.DATABASECHANGELOG' doesn't exist" when running migration on a fresh database

### ISSUE THREAD

**Issue by wololock** [micronaut-projects/micronaut-liquibase#565]
### Expected Behavior

Executing Liquibase migration scripts using `micronaut-jdbc-tomcat` on a fresh database succeeds with no error.

### Actual Behaviour

We noticed this issue while upgrading one of our applications from Micronaut 4.3.8 to the latest Micronaut 4.6.1. We use `micronaut-liquibase` with the `micronaut-jdbc-tomcat` in the integration tests and it stopped working after upgrading to the latest Micronaut version (4.6.1). I want to report this issue to raise your awareness - I'm quite confident there's not much you can do about it since the root cause of this problem exists in the underlying libraries.

When we execute Liquibase migration scripts on a fresh database, it fails with the following errors:

```
Bean definition [javax.sql.DataSource] could not be loaded: Migration failed! Liquibase encountered an exception.
io.micronaut.context.exceptions.BeanInstantiationException: Bean definition [javax.sql.DataSource] could not be loaded: Migration failed! Liquibase encountered an exception.
	at io.micronaut.context.DefaultBeanContext.initializeContext(DefaultBeanContext.java:1996)
	at io.micronaut.context.DefaultApplicationContext.initializeContext(DefaultApplicationContext.java:307)
	at io.micronaut.context.DefaultBeanContext.configureAndStartContext(DefaultBeanContext.java:3314)
	at io.micronaut.context.DefaultBeanContext.start(DefaultBeanContext.java:345)
	at io.micronaut.context.DefaultApplicationContext.start(DefaultApplicationContext.java:215)
	at io.micronaut.test.extensions.AbstractMicronautExtension.startApplicationContext(AbstractMicronautExtension.java:507)
	at io.micronaut.test.extensions.spock.MicronautSpockExtension.startApplicationContext(MicronautSpockExtension.java:299)
	at io.micronaut.test.extensions.AbstractMicronautExtension.beforeClass(AbstractMicronautExtension.java:346)
	at io.micronaut.test.extensions.spock.MicronautSpockExtension.lambda$visitSpecAnnotation$2(MicronautSpockExtension.java:93)
	at org.spockframework.runtime.extension.MethodInvocation.proceed(MethodInvocation.java:101)
	at org.spockframework.runtime.model.MethodInfo.invoke(MethodInfo.java:156)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
Caused by: io.micronaut.runtime.exceptions.ApplicationStartupException: Migration failed! Liquibase encountered an exception.
	at io.micronaut.liquibase.AbstractLiquibaseMigration.migrate(AbstractLiquibaseMigration.java:146)
	at io.micronaut.liquibase.AbstractLiquibaseMigration.forceRun(AbstractLiquibaseMigration.java:100)
	at io.micronaut.liquibase.AbstractLiquibaseMigration.run(AbstractLiquibaseMigration.java:86)
	at io.micronaut.liquibase.LiquibaseMigrationRunner.lambda$onCreated$1(LiquibaseMigrationRunner.java:64)
	at java.base/java.util.Optional.ifPresent(Optional.java:178)
	at io.micronaut.liquibase.LiquibaseMigrationRunner.onCreated(LiquibaseMigrationRunner.java:62)
	at io.micronaut.liquibase.LiquibaseMigrationRunner.onCreated(LiquibaseMigrationRunner.java:39)
	at io.micronaut.context.DefaultBeanContext.triggerBeanCreatedEventListener(DefaultBeanContext.java:2383)
	at io.micronaut.context.DefaultBeanContext.postBeanCreated(DefaultBeanContext.java:2360)
	at io.micronaut.context.DefaultBeanContext.doCreateBean(DefaultBeanContext.java:2304)
	at io.micronaut.context.DefaultBeanContext.doCreateBean(DefaultBeanContext.java:2312)
	at io.micronaut.context.DefaultBeanContext.createRegistration(DefaultBeanContext.java:3123)
	at io.micronaut.context.SingletonScope.getOrCreate(SingletonScope.java:80)
	at io.micronaut.context.DefaultBeanContext.findOrCreateSingletonBeanRegistration(DefaultBeanContext.java:3025)
	at io.micronaut.context.DefaultBeanContext.initializeEagerBean(DefaultBeanContext.java:2689)
	at io.micronaut.context.DefaultBeanContext.initializeContext(DefaultBeanContext.java:1990)
	... 11 more
Caused by: liquibase.exception.LiquibaseException: java.lang.reflect.UndeclaredThrowableException
	at liquibase.Liquibase.runInScope(Liquibase.java:1292)
	at liquibase.Liquibase.visitInScope(Liquibase.java:1108)
	at liquibase.Liquibase.listUnrunChangeSets(Liquibase.java:1073)
	at io.micronaut.liquibase.AbstractLiquibaseMigration.isUpdateNeeded(AbstractLiquibaseMigration.java:325)
	at io.micronaut.liquibase.AbstractLiquibaseMigration.performUpdateIfNeeded(AbstractLiquibaseMigration.java:183)
	at io.micronaut.liquibase.AbstractLiquibaseMigration.migrate(AbstractLiquibaseMigration.java:140)
	... 26 more
Caused by: java.lang.reflect.UndeclaredThrowableException
	at liquibase.executor.jvm.JdbcExecutor.query(JdbcExecutor.java:220)
	at liquibase.executor.jvm.JdbcExecutor.query(JdbcExecutor.java:238)
	at liquibase.executor.jvm.JdbcExecutor.queryForObject(JdbcExecutor.java:246)
	at liquibase.executor.jvm.JdbcExecutor.queryForObject(JdbcExecutor.java:261)
	at liquibase.executor.jvm.JdbcExecutor.queryForInt(JdbcExecutor.java:282)
	at liquibase.executor.jvm.JdbcExecutor.queryForInt(JdbcExecutor.java:277)
	at liquibase.snapshot.SnapshotGeneratorFactory.has(SnapshotGeneratorFactory.java:110)
	at liquibase.snapshot.SnapshotGeneratorFactory.hasDatabaseChangeLogTable(SnapshotGeneratorFactory.java:282)
	at liquibase.changelog.StandardChangeLogHistoryService.hasDatabaseChangeLogTable(StandardChangeLogHistoryService.java:86)
	at liquibase.changelog.StandardChangeLogHistoryService.getRanChangeSets(StandardChangeLogHistoryService.java:309)
	at liquibase.database.AbstractJdbcDatabase.getRanChangeSetList(AbstractJdbcDatabase.java:1070)
	at liquibase.changelog.DatabaseChangeLog.validate(DatabaseChangeLog.java:379)
	at liquibase.Liquibase.lambda$visitInScope$20(Liquibase.java:1113)
	at liquibase.Scope.lambda$child$0(Scope.java:191)
	at liquibase.Scope.child(Scope.java:200)
	at liquibase.Scope.child(Scope.java:190)
	at liquibase.Scope.child(Scope.java:169)
	at liquibase.Liquibase.runInScope(Liquibase.java:1290)
	... 31 more
Caused by: java.lang.reflect.InvocationTargetException
	at org.apache.tomcat.jdbc.pool.StatementFacade$StatementProxy.invoke(StatementFacade.java:130)
	... 49 more
Caused by: java.sql.SQLSyntaxErrorException: Table 'test.DATABASECHANGELOG' doesn't exist
	at com.mysql.cj.jdbc.exceptions.SQLError.createSQLException(SQLError.java:112)
	at com.mysql.cj.jdbc.exceptions.SQLExceptionsMapping.translateException(SQLExceptionsMapping.java:113)
	at com.mysql.cj.jdbc.ClientPreparedStatement.executeInternal(ClientPreparedStatement.java:938)
	at com.mysql.cj.jdbc.ClientPreparedStatement.executeQuery(ClientPreparedStatement.java:1004)
	... 50 more


Migration failed! Liquibase encountered an exception.
io.micronaut.runtime.exceptions.ApplicationStartupException: Migration failed! Liquibase encountered an exception.
	at app//io.micronaut.liquibase.AbstractLiquibaseMigration.migrate(AbstractLiquibaseMigration.java:146)
	at app//io.micronaut.liquibase.AbstractLiquibaseMigration.forceRun(AbstractLiquibaseMigration.java:100)
	at app//io.micronaut.liquibase.AbstractLiquibaseMigration.run(AbstractLiquibaseMigration.java:86)
	at app//io.micronaut.liquibase.LiquibaseMigrationRunner.lambda$onCreated$1(LiquibaseMigrationRunner.java:64)
	at java.base@21.0.2/java.util.Optional.ifPresent(Optional.java:178)
	at app//io.micronaut.liquibase.LiquibaseMigrationRunner.onCreated(LiquibaseMigrationRunner.java:62)
	at app//io.micronaut.liquibase.LiquibaseMigrationRunner.onCreated(LiquibaseMigrationRunner.java:39)
	at app//io.micronaut.context.DefaultBeanContext.triggerBeanCreatedEventListener(DefaultBeanContext.java:2383)
	at app//io.micronaut.context.DefaultBeanContext.postBeanCreated(DefaultBeanContext.java:2360)
	at app//io.micronaut.context.DefaultBeanContext.doCreateBean(DefaultBeanContext.java:2304)
	at app//io.micronaut.context.DefaultBeanContext.doCreateBean(DefaultBeanContext.java:2312)
	at app//io.micronaut.context.DefaultBeanContext.createRegistration(DefaultBeanContext.java:3123)
	at app//io.micronaut.context.SingletonScope.getOrCreate(SingletonScope.java:80)
	at app//io.micronaut.context.DefaultBeanContext.findOrCreateSingletonBeanRegistration(DefaultBeanContext.java:3025)
	at app//io.micronaut.context.DefaultBeanContext.initializeEagerBean(DefaultBeanContext.java:2689)
	at app//io.micronaut.context.DefaultBeanContext.initializeContext(DefaultBeanContext.java:1990)
	at app//io.micronaut.context.DefaultApplicationContext.initializeContext(DefaultApplicationContext.java:307)
	at app//io.micronaut.context.DefaultBeanContext.configureAndStartContext(DefaultBeanContext.java:3314)
	at app//io.micronaut.context.DefaultBeanContext.start(DefaultBeanContext.java:345)
	at app//io.micronaut.context.DefaultApplicationContext.start(DefaultApplicationContext.java:215)
	at app//io.micronaut.test.extensions.AbstractMicronautExtension.startApplicationContext(AbstractMicronautExtension.java:507)
	at app//io.micronaut.test.extensions.spock.MicronautSpockExtension.startApplicationContext(MicronautSpockExtension.java:299)
	at app//io.micronaut.test.extensions.AbstractMicronautExtension.beforeClass(AbstractMicronautExtension.java:346)
	at app//io.micronaut.test.extensions.spock.MicronautSpockExtension.lambda$visitSpecAnnotation$2(MicronautSpockExtension.java:93)
	at app//org.spockframework.runtime.extension.MethodInvocation.proceed(MethodInvocation.java:101)
	at app//org.spockframework.runtime.model.MethodInfo.invoke(MethodInfo.java:156)
	at java.base@21.0.2/java.util.ArrayList.forEach(ArrayList.java:1596)
Caused by: liquibase.exception.LiquibaseException: java.lang.reflect.UndeclaredThrowableException
	at app//liquibase.Liquibase.runInScope(Liquibase.java:1292)
	at app//liquibase.Liquibase.visitInScope(Liquibase.java:1108)
	at app//liquibase.Liquibase.listUnrunChangeSets(Liquibase.java:1073)
	at app//io.micronaut.liquibase.AbstractLiquibaseMigration.isUpdateNeeded(AbstractLiquibaseMigration.java:325)
	at app//io.micronaut.liquibase.AbstractLiquibaseMigration.performUpdateIfNeeded(AbstractLiquibaseMigration.java:183)
	at app//io.micronaut.liquibase.AbstractLiquibaseMigration.migrate(AbstractLiquibaseMigration.java:140)
	... 26 more
Caused by: java.lang.reflect.UndeclaredThrowableException
	at liquibase.executor.jvm.JdbcExecutor.query(JdbcExecutor.java:220)
	at liquibase.executor.jvm.JdbcExecutor.query(JdbcExecutor.java:238)
	at liquibase.executor.jvm.JdbcExecutor.queryForObject(JdbcExecutor.java:246)
	at liquibase.executor.jvm.JdbcExecutor.queryForObject(JdbcExecutor.java:261)
	at liquibase.executor.jvm.JdbcExecutor.queryForInt(JdbcExecutor.java:282)
	at liquibase.executor.jvm.JdbcExecutor.queryForInt(JdbcExecutor.java:277)
	at liquibase.snapshot.SnapshotGeneratorFactory.has(SnapshotGeneratorFactory.java:110)
	at liquibase.snapshot.SnapshotGeneratorFactory.hasDatabaseChangeLogTable(SnapshotGeneratorFactory.java:282)
	at liquibase.changelog.StandardChangeLogHistoryService.hasDatabaseChangeLogTable(StandardChangeLogHistoryService.java:86)
	at liquibase.changelog.StandardChangeLogHistoryService.getRanChangeSets(StandardChangeLogHistoryService.java:309)
	at liquibase.database.AbstractJdbcDatabase.getRanChangeSetList(AbstractJdbcDatabase.java:1070)
	at liquibase.changelog.DatabaseChangeLog.validate(DatabaseChangeLog.java:379)
	at liquibase.Liquibase.lambda$visitInScope$20(Liquibase.java:1113)
	at liquibase.Scope.lambda$child$0(Scope.java:191)
	at liquibase.Scope.child(Scope.java:200)
	at liquibase.Scope.child(Scope.java:190)
	at liquibase.Scope.child(Scope.java:169)
	at liquibase.Liquibase.runInScope(Liquibase.java:1290)
	... 31 more
Caused by: java.lang.reflect.InvocationTargetException
	at org.apache.tomcat.jdbc.pool.StatementFacade$StatementProxy.invoke(StatementFacade.java:130)
	... 49 more
Caused by: java.sql.SQLSyntaxErrorException: Table 'test.DATABASECHANGELOG' doesn't exist
	at com.mysql.cj.jdbc.exceptions.SQLError.createSQLException(SQLError.java:112)
	at com.mysql.cj.jdbc.exceptions.SQLExceptionsMapping.translateException(SQLExceptionsMapping.java:113)
	at com.mysql.cj.jdbc.ClientPreparedStatement.executeInternal(ClientPreparedStatement.java:938)
	at com.mysql.cj.jdbc.ClientPreparedStatement.executeQuery(ClientPreparedStatement.java:1004)
	... 50 more
```

The root cause of this issue was introduced by `liquibase-core` version `4.29.0` and the following change made in the `SnapshotFactoryGenerator` class:

![image](https://github.com/user-attachments/assets/eb09b1e6-649b-48f3-86d4-ec685f8a9718)

Source: https://github.com/liquibase/liquibase/compare/v4.28.0...v4.29.0#diff-0e46f8aeb1afb5d682308e3271825803d6eccf5e38af69a9a034b2da01f546b5

Up until version `4.29.0`, the implementation of `has(...)` method used a raw SQL query that threw a proper (and expected) `SQLException` when a given table didn't exist. This behavior changed starting from `liquibase-core:4.29.0` when they switched to the prepared statement - it turns out that Tomcat JDBC connection throws a `java.lang.reflect.InvocationTargetException` that is not caught by try-catch that expects an exception from the `SQLException` family.

![Screenshot from 2024-09-13 12-53-52](https://github.com/user-attachments/assets/dbff9c6f-45ef-414a-9abe-afbd40e8b119)

When I replace `micronaut-jdbc-tomcat` with e.g. `micronaut-jdbc-hikari`, this problem does not exist. In case of Hikari, a `HikariProxyPreparedStatement` object throws `SQLSyntaxErrorException` when the prepared statement query is executed against a table that does not exist. 

It's worth mentioning that switching from `micronaut-jdbc-tomcat` to something like `micronaut-jdbc-hikari` solves the problem. I am also aware that this problem is not generated directly by `micronaut-liquibase` code, but I thought it's worth to report this issue so that other Micronaut users may find it if they face a similar issue.

### Steps To Reproduce

1. Create a new Micronaut 4.6.x application using Micronaut Launcher with `liquibase` and `mysql` features selected.
2. Replace `micronaut-jdbc-hikari` with `micronaut-jdbc-tomcat` in the `build.gradle`.
3. Execute generated unit test that will run Liquibase migrations.

### Example Application

https://github.com/wololock/micronaut-liquibase-demo

### Version

4.6.1

**Comment by radovanradic** [comment]
This should be fixed in 4.6.2 (changed dependencies to use tomcat-jdbc 10.1.29 that resolves this issue)
There was similar ticket for this https://github.com/micronaut-projects/micronaut-core/issues/11178

**Comment by wololock** [comment]
Thank you @radovanradic - I can confirm that upgrading to `4.6.2` fixes this problem :+1: 

![Screenshot from 2024-09-13 17-09-40](https://github.com/user-attachments/assets/d0bce8a7-7116-447b-a676-74bdb1a0f5f9)
